File: labs-opensource/M03B-context-engineering/solution/context_budget.js
/**
 * M03B Lab: The ContextBudget Class — SOLUTION
 * =============================================
 * Run the full lab: copy this over starter/context_budget.js, then node diagnose.js
 */

import { getEncoding } from "js-tiktoken";
import OpenAI from "openai";

// cl100k_base is OpenAI's tokenizer — a good approximation for Mistral/Llama
const enc = getEncoding("cl100k_base");

/** Count tokens in a string or JSON-serializable object. */
export function countTokens(text) {
  const s = typeof text === "string" ? text : JSON.stringify(text);
  return Math.max(1, enc.encode(s).length);
}

export const MODEL_LIMITS = {
  mistral: 32_768,
  mixtral: 32_768,
  llama3: 131_072,
  gemma2: 8_192,
};

const SUMMARY_PROMPT = `Summarize this conversation in 3-5 sentences.
CRITICAL: preserve every order ID, customer ID, tracking number, exact dollar amount,
and explicit user decision. Drop only retry errors, duplicate results, and resolved detours.

Conversation:
{transcript}

Summary:`;

/**
 * Summarize older turns; keep the last `keepRecent` turns verbatim.
 * GOTCHA: this is a SEPARATE model call — never ask the main model to
 * summarize the context it is currently reading.
 */
export async function summarizeHistory(history, keepRecent = 4) {
  if (history.length <= keepRecent) return history;

  const older = history.slice(0, -keepRecent);
  const recent = history.slice(-keepRecent);
  const transcript = older.map((m) => `${m.role}: ${m.content}`).join("\n");

  const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });
  try {
    const result = await client.chat.completions.create({
      model: "mistral",
      messages: [{ role: "user", content: SUMMARY_PROMPT.replace("{transcript}", transcript) }],
      max_tokens: 250,
    });
    const summaryText = result.choices[0].message.content ?? "";
    return [
      { role: "user", content: `[Summary of ${older.length} earlier turns] ${summaryText}` },
      ...recent,
    ];
  } catch (e) {
    console.error(`Summarization failed: ${e.message}; falling back to truncation.`);
    return recent;
  }
}

export class ContextBudget {
  constructor(opts = {}) {
    this.systemPrompt = opts.systemPrompt ?? "";
    this.toolDefinitions = opts.toolDefinitions ?? [];
    this.retrievedDocs = opts.retrievedDocs ?? [];
    this.history = opts.history ?? [];
    this.toolResults = opts.toolResults ?? [];
    this.currentUserMessage = opts.currentUserMessage ?? "";
    this.model = opts.model ?? "mistral";
    this.reserveOutput = opts.reserveOutput ?? 2048;
  }

  get maxTokens() {
    return (MODEL_LIMITS[this.model] ?? 32_768) - this.reserveOutput;
  }

  /** Per-layer token breakdown. */
  account() {
    return {
      system: countTokens(this.systemPrompt),
      tools: this.toolDefinitions.length ? countTokens(this.toolDefinitions) : 0,
      retrieved: this.retrievedDocs.reduce((s, d) => s + countTokens(d), 0),
      history: this.history.reduce((s, m) => s + countTokens(m.content), 0),
      toolResults: this.toolResults.reduce((s, r) => s + countTokens(r), 0),
      current: countTokens(this.currentUserMessage),
    };
  }

  total() {
    return Object.values(this.account()).reduce((a, b) => a + b, 0);
  }

  remaining() {
    return this.maxTokens - this.total();
  }

  fits() {
    return this.total() <= this.maxTokens;
  }

  /** Return recommended strategy based on budget utilization. */
  strategy() {
    const utilization = this.total() / this.maxTokens;
    if (utilization < 0.60) return "ok";        // plenty of room
    if (utilization < 0.75) return "compress";  // crop tool defs, compress results
    if (utilization < 0.90) return "summarize"; // summarize old history turns
    return "critical";                          // system + last 2 turns + current only
  }

  /** Apply the recommended strategy in place. Returns this for chaining. */
  crop() {
    const strat = this.strategy();
    if (strat === "ok") return this;

    if (["compress", "summarize", "critical"].includes(strat)) {
      // Crop tool definitions when in trouble (saves ~900 tok for 4 tools)
      this.toolDefinitions = [];
    }

    if (["summarize", "critical"].includes(strat)) {
      const keep = strat === "critical" ? 2 : 4;
      this.history = this.history.slice(-keep);
      this.toolResults = this.toolResults.slice(-2);
    }

    if (strat === "critical") {
      // Nuclear option: drop all retrieved docs too
      this.retrievedDocs = [];
    }

    return this;
  }

  /** Summarize old history turns. Call when strategy() === "summarize". */
  async checkpoint(keepRecent = 4) {
    this.history = await summarizeHistory(this.history, keepRecent);
    return this;
  }

  /**
   * Assemble { system, messages } for client.chat.completions.create().
   * Retrieved docs are appended to the CURRENT message, not the system
   * prompt — end position = maximum recall on Mistral-7B.
   */
  buildMessages() {
    let retrievedBlock = "";
    if (this.retrievedDocs.length) {
      retrievedBlock =
        "\n\n<reference_docs>\n" + this.retrievedDocs.join("\n---\n") + "\n</reference_docs>\n\n";
    }

    const messages = [
      ...this.history,
      { role: "user", content: retrievedBlock + this.currentUserMessage },
    ];
    return { system: this.systemPrompt, messages };
  }
}

File: labs-opensource/M03B-context-engineering/solution/context_budget.py
"""
M03B Lab: The ContextBudget Class — SOLUTION
=============================================
Run the self-test: python context_budget.py
Run the full lab:  copy this over starter/context_budget.py, then python diagnose.py
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field

import tiktoken
from openai import OpenAI

# cl100k_base is OpenAI's tokenizer — a good approximation for Mistral/Llama
_enc = tiktoken.get_encoding("cl100k_base")


def count_tokens(text) -> int:
    """Count tokens in a string or JSON-serializable object."""
    if not isinstance(text, str):
        text = json.dumps(text, separators=(",", ":"))
    return max(1, len(_enc.encode(text)))


MODEL_LIMITS = {
    "mistral": 32_768,
    "mixtral": 32_768,
    "llama3": 131_072,
    "gemma2": 8_192,
}

SUMMARY_PROMPT = """Summarize this conversation in 3-5 sentences.
CRITICAL: preserve every order ID, customer ID, tracking number, exact dollar amount,
and explicit user decision. Drop only retry errors, duplicate results, and resolved detours.

Conversation:
{transcript}

Summary:"""


def summarize_history(history: list[dict], keep_recent: int = 4) -> list[dict]:
    """Summarize older turns; keep the last `keep_recent` turns verbatim.

    GOTCHA: this is a SEPARATE model call — never ask the main model to
    summarize the context it is currently reading.
    """
    if len(history) <= keep_recent:
        return history

    older = history[:-keep_recent]
    recent = history[-keep_recent:]
    transcript = "\n".join(f"{m['role']}: {m['content']}" for m in older)

    client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")
    try:
        result = client.chat.completions.create(
            model="mistral",
            messages=[{"role": "user", "content": SUMMARY_PROMPT.format(transcript=transcript)}],
            max_tokens=250,
        )
        summary_text = result.choices[0].message.content or ""
    except Exception as e:
        print(f"Summarization failed: {e}; falling back to truncation.")
        return recent

    summary_msg = {
        "role": "user",
        "content": f"[Summary of {len(older)} earlier turns] {summary_text}",
    }
    return [summary_msg] + recent


@dataclass
class ContextBudget:
    """Accounts for and curates the six layers of a local model context."""

    system_prompt: str = ""
    tool_definitions: list = field(default_factory=list)
    retrieved_docs: list[str] = field(default_factory=list)
    history: list[dict] = field(default_factory=list)
    tool_results: list[str] = field(default_factory=list)
    current_user_message: str = ""
    model: str = "mistral"
    reserve_output: int = 2048

    @property
    def max_tokens(self) -> int:
        return MODEL_LIMITS.get(self.model, 32_768) - self.reserve_output

    def account(self) -> dict[str, int]:
        """Per-layer token breakdown."""
        return {
            "system": count_tokens(self.system_prompt),
            "tools": count_tokens(self.tool_definitions) if self.tool_definitions else 0,
            "retrieved": sum(count_tokens(d) for d in self.retrieved_docs),
            "history": sum(count_tokens(m["content"]) for m in self.history),
            "tool_results": sum(count_tokens(r) for r in self.tool_results),
            "current": count_tokens(self.current_user_message),
        }

    def total(self) -> int:
        return sum(self.account().values())

    def remaining(self) -> int:
        return self.max_tokens - self.total()

    def fits(self) -> bool:
        return self.total() <= self.max_tokens

    def strategy(self) -> str:
        """Return recommended strategy based on budget utilization."""
        utilization = self.total() / self.max_tokens
        if utilization < 0.60:
            return "ok"           # plenty of room — no action needed
        elif utilization < 0.75:
            return "compress"     # crop tool definitions, compress tool results
        elif utilization < 0.90:
            return "summarize"    # summarize old history turns
        else:
            return "critical"     # emergency: system + last 2 turns + current only

    def crop(self) -> "ContextBudget":
        """Apply the recommended strategy in place. Returns self for chaining."""
        strat = self.strategy()
        if strat == "ok":
            return self

        if strat in ("compress", "summarize", "critical"):
            # Crop tool definitions when in trouble (saves ~900 tok for 4 tools)
            self.tool_definitions = []

        if strat in ("summarize", "critical"):
            keep = 2 if strat == "critical" else 4
            self.history = self.history[-keep:]
            self.tool_results = self.tool_results[-2:]

        if strat == "critical":
            # Nuclear option: drop all retrieved docs too
            self.retrieved_docs = []

        return self

    def checkpoint(self, keep_recent: int = 4) -> "ContextBudget":
        """Summarize old history turns. Call when strategy() == 'summarize'."""
        self.history = summarize_history(self.history, keep_recent=keep_recent)
        return self

    def build_messages(self) -> tuple[str, list[dict]]:
        """Assemble (system, messages) for client.chat.completions.create().

        Retrieved docs are appended to the CURRENT message, not the system
        prompt — end position = maximum recall on Mistral-7B.
        """
        retrieved_block = ""
        if self.retrieved_docs:
            retrieved_block = (
                "\n\n<reference_docs>\n"
                + "\n---\n".join(self.retrieved_docs)
                + "\n</reference_docs>\n\n"
            )

        messages = list(self.history)
        messages.append({
            "role": "user",
            "content": retrieved_block + self.current_user_message,
        })
        return self.system_prompt, messages


if __name__ == "__main__":
    budget = ContextBudget(
        model="mistral",
        system_prompt="You are an order-tracking assistant. Always cite order IDs.",
        current_user_message="Where is my order ORD-88421?",
        retrieved_docs=["ORD-88421: in transit, estimated delivery Nov 8 via FedEx."],
    )
    print(f"Tokens: {budget.total()} / {budget.max_tokens} ({budget.strategy()})")
    system, messages = budget.build_messages()
    print(f"Messages: {len(messages)}, last content: {messages[-1]['content'][:80]}")
